### PR TITLE
impl(generator): interrogate existing protos for field number info

### PR DESCRIPTION
**Pull request by scotthart**
part of the work for #11095

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071)

### CODE REVIEW

File: generator/internal/discovery_type_vertex_test.cc [hunk 1]
```diff
@@ -673,6 +675,66 @@ message TestSchema {
...
+  auto const* file_descriptor = pool().FindFileByName("test.proto");
+  ASSERT_THAT(file_descriptor, NotNull());
+
+  auto const* message_descriptor =
+      file_descriptor->FindMessageTypeByName("FieldsOnly");
+  ASSERT_THAT(message_descriptor, NotNull());
+  auto message_properties =
+      DiscoveryTypeVertex::DetermineReservedAndMaxFieldNumbers(
+          *message_descriptor);
+  EXPECT_THAT(message_properties.max_field_number, Eq(2));
```
Comment (devbww) [inline generator/internal/discovery_type_vertex_test.cc:716]: Expectations for `reserved_numbers`?

File: generator/internal/discovery_type_vertex.h [hunk 2]
```diff
@@ -70,6 +71,16 @@ class DiscoveryTypeVertex {
   static StatusOr<TypeInfo> DetermineTypeAndSynthesis(
       nlohmann::json const& v, std::string const& field_name);
 
+  struct MessageProperties {
+    std::set<int> reserved_numbers;
+    int max_field_number;
+  };
+
+  // Examines the message Descriptor to determined the reserved field numbers
+  // and the max field number either in use or reserved.
```
Comment (devbww) [inline generator/internal/discovery_type_vertex.h:80]: Q.  What does `max_field_number` buy you?  If 42 was the largest field used, but 12 was unused/unreserved, would we not want to use 12?

File: generator/internal/discovery_type_vertex.cc [hunk 3]
```diff
@@ -183,6 +183,31 @@ StatusOr<std::string> DiscoveryTypeVertex::FormatMessage(
...
+DiscoveryTypeVertex::DetermineReservedAndMaxFieldNumbers(
+    google::protobuf::Descriptor const& message_descriptor) {
+  MessageProperties message_properties{{}, 0};
+  for (auto r = 0; r != message_descriptor.reserved_range_count(); ++r) {
+    auto const* reserved_range = message_descriptor.reserved_range(r);
+    for (int j = reserved_range->start; j <= reserved_range->end; ++j) {
+      message_properties.reserved_numbers.insert(j);
+    }
+    // google::protobuf::ReservedRange.end returns the next available value, not
+    // the actual end reserved value.
```
Comment (devbww) [inline generator/internal/discovery_type_vertex.cc:196]: Perhaps "next available" would be a better interface than "max used" here too.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071)* status: 0 of 3 files reviewed, 3 unresolved discussions (waiting on @devbww)

___
*[`generator/internal/discovery_type_vertex.h` line 80 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071#gh-1260056128:-N_51pIS5jGrr5cUQel-:b-ax1jbb) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5b54cbb7dd1375dce795ecb8a61246031c6ca584/generator/internal/discovery_type_vertex.h#L80)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Q.  What does `max_field_number` buy you?  If 42 was the largest field used, but 12 was unused/unreserved, would we not want to use 12?
</blockquote></details>

The protos in question will always be generated such that there are no gaps in the field numbers. Every field number less than the max will either be in use or reserved.
___
*[`generator/internal/discovery_type_vertex.cc` line 196 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071#gh-1260059174:-N_527Nm7PY-h0TB_6PX:b-yyiv2f) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5b54cbb7dd1375dce795ecb8a61246031c6ca584/generator/internal/discovery_type_vertex.cc#L196)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Perhaps "next available" would be a better interface than "max used" here too.
</blockquote></details>

I think it's a case of 6 or half a dozen. If I return the max and need next available, I add 1. If I return next available, but need max, then I subtract 1.
___
*[`generator/internal/discovery_type_vertex_test.cc` line 716 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071#gh-1260044027:-N_52JZNBV2X0fG_F-L1:b307z23) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5b54cbb7dd1375dce795ecb8a61246031c6ca584/generator/internal/discovery_type_vertex_test.cc#L716)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Expectations for `reserved_numbers`?
</blockquote></details>

Oops; added.

**Review by devbww — COMMENTED** [review]
> Reviewed 1 of 3 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071)* status: all files reviewed, 2 unresolved discussions (waiting on @scotthart)

___
*[`generator/internal/discovery_type_vertex.h` line 80 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071#gh-1260056128:-N_56TVO3mmFr9YXisB2:b-32qpb4) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5b54cbb7dd1375dce795ecb8a61246031c6ca584/generator/internal/discovery_type_vertex.h#L80)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

The protos in question will always be generated such that there are no gaps in the field numbers. Every field number less than the max will either be in use or reserved.
</blockquote></details>

Then what does `reserved_numbers` buy you?  Just record the max/next-avail.
___
*[`generator/internal/discovery_type_vertex.cc` line 196 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071#gh-1260059174:-N_55IDP9l3POlVQROw7:bs8h4y) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5b54cbb7dd1375dce795ecb8a61246031c6ca584/generator/internal/discovery_type_vertex.cc#L196)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

I think it's a case of 6 or half a dozen. If I return the max and need next available, I add 1. If I return next available, but need max, then I subtract 1.
</blockquote></details>

I guess I wonder when you'll need max.

"next available" also (1) avoids the question of the semantics of "max used" when none are used, and (2) continues the ubiquitous open-ended-interval model of C++.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @devbww)

___
*[`generator/internal/discovery_type_vertex.h` line 80 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071#gh-1260056128:-N_5FN402UQD5BxZfHRa:b-jpb8ki) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5b54cbb7dd1375dce795ecb8a61246031c6ca584/generator/internal/discovery_type_vertex.h#L80)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Then what does `reserved_numbers` buy you?  Just record the max/next-avail.
</blockquote></details>

I need to know the current reserved_numbers such that I can emit them again when writing the next generation of the same proto message.
___
*[`generator/internal/discovery_type_vertex.cc` line 196 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071#gh-1260059174:-N_5Fb8XD5x5Pp6UiLs0:b-4h7k7j) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5b54cbb7dd1375dce795ecb8a61246031c6ca584/generator/internal/discovery_type_vertex.cc#L196)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

I guess I wonder when you'll need max.

"next available" also (1) avoids the question of the semantics of "max used" when none are used, and (2) continues the ubiquitous open-ended-interval model of C++.
</blockquote></details>

changed to next_available

**Review by devbww — APPROVED** [review]
> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071)* status: all files reviewed, 1 unresolved discussion (waiting on @scotthart)

___
*[`generator/internal/discovery_type_vertex.h` line 80 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071#gh-1260056128:-N_5a3Fl550wJAmLVbpv:b50xhff) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5b54cbb7dd1375dce795ecb8a61246031c6ca584/generator/internal/discovery_type_vertex.h#L80)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

I need to know the current reserved_numbers such that I can emit them again when writing the next generation of the same proto message.
</blockquote></details>

OK.  Although I presume you need to know a lot more than the `MessageProperties` when writing the next generation, like all the actual field types/names/tags.  I guess these properties just get moved to the code that knows those other things, but I would also have guessed that they would have been more co-located than a static function indicates.
___
*[`generator/internal/discovery_type_vertex_test.cc` line 704 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071#-N_5cj86AjnhdFwx8p2z:-N_5cj86AjnhdFwx8p3-:b-qei0u3) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/15d0ff059c486249ae647f6818b6d302e2991eec/generator/internal/discovery_type_vertex_test.cc#L704)):*
> Every field number less than the max will either be in use or reserved.

These messages don't appear to meet those expectations.  Is that (and its non-diagnosis) a feature or a bug?

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071)* status: all files reviewed, 1 unresolved discussion (waiting on @devbww)

___
*[`generator/internal/discovery_type_vertex.h` line 80 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071#gh-1260056128:-N_9eQxo3mZlS-U-oTl9:b-x0uc49) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5b54cbb7dd1375dce795ecb8a61246031c6ca584/generator/internal/discovery_type_vertex.h#L80)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

OK.  Although I presume you need to know a lot more than the `MessageProperties` when writing the next generation, like all the actual field types/names/tags.  I guess these properties just get moved to the code that knows those other things, but I would also have guessed that they would have been more co-located than a static function indicates.
</blockquote></details>

There will be many more code changes in the near future.
___
*[`generator/internal/discovery_type_vertex_test.cc` line 704 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12071#-N_5cj86AjnhdFwx8p2z:-N_9eXBzCDzIHaskjVBy:bi20ct9) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/15d0ff059c486249ae647f6818b6d302e2991eec/generator/internal/discovery_type_vertex_test.cc#L704)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

> Every field number less than the max will either be in use or reserved.

These messages don't appear to meet those expectations.  Is that (and its non-diagnosis) a feature or a bug?
</blockquote></details>

These tests were only focused on determining the max/next_available field number and identifying all reserved field numbers, not ensuring that an invariant exists on the output from some other piece of code.

**Pull request merged**